src/lib/Gen/hbs/HandlebarsGenerator.cpp

```diff
@@ -15,13 +15,73 @@
 #include "Builder.hpp"
 #include "MultiPageVisitor.hpp"
 #include "SinglePageVisitor.hpp"
+
+#include "lib/Lib/TagfileWriter.hpp"
+#include "lib/Support/RawOstream.hpp"
+
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+
 #include <mrdocs/Support/Path.hpp>
+
+#include <fstream>
 #include <sstream>
 
 namespace clang {
 namespace mrdocs {
 namespace hbs {
 
+namespace {
+
+/** Generate a Tagfile for cross-referencing. */
+Error
+generateTagfile(
+    HandlebarsCorpus const& hbsCorpus,
+    std::string_view tagFilePath,
+    std::string_view generatedFilesPath,
+    bool isSinglePage)
+{
+    if (tagFilePath.empty())
+        return Error::success();
+
+    auto tagFileWriterEx = TagfileWriter::create(hbsCorpus, tagFilePath, generatedFilesPath, isSinglePage);
+    if (!tagFileWriterEx)
+        return tagFileWriterEx.error();
+
+    auto tagfileWriter = std::move(*tagFileWriterEx);
+    tagfileWriter.initialize();
+    visit(hbsCorpus->globalNamespace(), tagfileWriter);
+    tagfileWriter.finalize();
+    return Error::success();
+}
+
+std::string 
+getSinglePageFileName(
+    std::string_view outputPath, 
+    std::string_view extension)
+{
+    namespace path = llvm::sys::path;
+
+    using SmallString = llvm::SmallString<0>;
+
+    SmallString ext(".");
+    ext += extension;
+
+    SmallString fileName(outputPath);
+    if(path::extension(outputPath).compare_insensitive(ext) != 0)
+    {
+        // directory specified
```

Comment:
So, if the outputPath doesn't end in `.html` or `.adoc` (let's say "x.jpg"), do we consider this a directory?

And then we append "reference.html" or "reference.adoc" to it? What for?

Are you duplicating the code from BuildOne here? More duplicated code is everything we don't need. In the future, we'll never remember to replicate the changes in multiple places.

We can create a free function in Generator and call it from both places. We can also improve the logic of the free function (Vinnie wrote that code in Feb 2023 in a hurry) to check if the outputPath is not a directory and adjust it accordingly.

As for the new logic, we have kind of a conflict here. If output path is already a directory (`isDirectory(outputPath)`) then things are clear. If the output path doesn't exist yet, then we don't know if the user intended a directory where "reference.adoc/html" should be appended or just a file without an extension. 

If the directory doesn't exist yet, we could assume anything with a dot is a file and anything without a dot should be a directory. Or we could return an error. All I know is the current heuristic (the filename matching the extension of the generator) is not good because extensions change and if the output path is "index.htm", we know the user doesn't want "index.htm/reference.html".